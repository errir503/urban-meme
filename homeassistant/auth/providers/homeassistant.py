"""Home Assistant auth provider."""
from __future__ import annotations

import asyncio
import base64
from collections.abc import Mapping
import logging
from typing import Any, cast

import bcrypt
import voluptuous as vol

from homeassistant.const import CONF_ID
from homeassistant.core import HomeAssistant, callback
from homeassistant.data_entry_flow import FlowResult
from homeassistant.exceptions import HomeAssistantError

from . import AUTH_PROVIDER_SCHEMA, AUTH_PROVIDERS, AuthProvider, LoginFlow
from ..models import Credentials, UserMeta

STORAGE_VERSION = 1
STORAGE_KEY = "auth_provider.homeassistant"


def _disallow_id(conf: dict[str, Any]) -> dict[str, Any]:
    """Disallow ID in config."""
    if CONF_ID in conf:
        raise vol.Invalid("ID is not allowed for the homeassistant auth provider.")

    return conf


CONFIG_SCHEMA = vol.All(AUTH_PROVIDER_SCHEMA, _disallow_id)


@callback
def async_get_provider(hass: HomeAssistant) -> HassAuthProvider:
    """Get the provider."""
    for prv in hass.auth.auth_providers:
        if prv.type == "homeassistant":
            return cast(HassAuthProvider, prv)

    raise RuntimeError("Provider not found")


class InvalidAuth(HomeAssistantError):
    """Raised when we encounter invalid authentication."""


class InvalidUser(HomeAssistantError):
    """Raised when invalid user is specified.

    Will not be raised when validating authentication.
    """


class Data:
    """Hold the user data."""

    def __init__(self, hass: HomeAssistant) -> None:
        """Initialize the user data store."""
        self.hass = hass
        self._store = hass.helpers.storage.Store(
            STORAGE_VERSION, STORAGE_KEY, private=True, atomic_writes=True
        )
        self._data: dict[str, Any] | None = None
        # Legacy mode will allow usernames to start/end with whitespace
        # and will compare usernames case-insensitive.
        # Remove in 2020 or when we launch 1.0.
        self.is_legacy = False

    @callback
    def normalize_username(self, username: str) -> str:
        """Normalize a username based on the mode."""
        if self.is_legacy:
            return username

        return username.strip().casefold()

    async def async_load(self) -> None:
        """Load stored data."""
        if (data := await self._store.async_load()) is None:
            data = {"users": []}

        seen: set[str] = set()

        for user in data["users"]:
            username = user["username"]

            # check if we have duplicates
            if (folded := username.casefold()) in seen:
                self.is_legacy = True

                logging.getLogger(__name__).warning(
                    "Home Assistant auth provider is running in legacy mode "
                    "because we detected usernames that are case-insensitive"
                    "equivalent. Please change the username: '%s'.",
                    username,
                )

                break

            seen.add(folded)

            # check if we have unstripped usernames
            if username != username.strip():
                self.is_legacy = True

                logging.getLogger(__name__).warning(
                    "Home Assistant auth provider is running in legacy mode "
                    "because we detected usernames that start or end in a "
                    "space. Please change the username: '%s'.",
                    username,
                )

                break

        self._data = data

    @property
    def users(self) -> list[dict[str, str]]:
        """Return users."""
        return self._data["users"]  # type: ignore[index,no-any-return]

    def validate_login(self, username: str, password: str) -> None:
        """Validate a username and password.

        Raises InvalidAuth if auth invalid.
        """
        username = self.normalize_username(username)
        dummy = b"$2b$12$CiuFGszHx9eNHxPuQcwBWez4CwDTOcLTX5CbOpV6gef2nYuXkY7BO"
        found = None

        # Compare all users to avoid timing attacks.
        for user in self.users:
            if self.normalize_username(user["username"]) == username:
                found = user

        if found is None:
            # check a hash to make timing the same as if user was found
            bcrypt.checkpw(b"foo", dummy)
            raise InvalidAuth

        user_hash = base64.b64decode(found["password"])

        # bcrypt.checkpw is timing-safe
        if not bcrypt.checkpw(password.encode(), user_hash):
            raise InvalidAuth

    def hash_password(  # pylint: disable=no-self-use
        self, password: str, for_storage: bool = False
    ) -> bytes:
        """Encode a password."""
        hashed: bytes = bcrypt.hashpw(password.encode(), bcrypt.gensalt(rounds=12))

        if for_storage:
            hashed = base64.b64encode(hashed)
        return hashed

    def add_auth(self, username: str, password: str) -> None:
        """Add a new authenticated user/pass."""
        username = self.normalize_username(username)

        if any(
            self.normalize_username(user["username"]) == username for user in self.users
        ):
            raise InvalidUser

        self.users.append(
            {
                "username": username,
                "password": self.hash_password(password, True).decode(),
            }
        )

    @callback
    def async_remove_auth(self, username: str) -> None:
        """Remove authentication."""
        username = self.normalize_username(username)

        index = None
        for i, user in enumerate(self.users):
            if self.normalize_username(user["username"]) == username:
                index = i
                break

        if index is None:
            raise InvalidUser

        self.users.pop(index)

    def change_password(self, username: str, new_password: str) -> None:
        """Update the password.

        Raises InvalidUser if user cannot be found.
        """
        username = self.normalize_username(username)

        for user in self.users:
            if self.normalize_username(user["username"]) == username:
                user["password"] = self.hash_password(new_password, True).decode()
                break
        else:
            raise InvalidUser

    async def async_save(self) -> None:
        """Save data."""
        await self._store.async_save(self._data)


@AUTH_PROVIDERS.register("homeassistant")
class HassAuthProvider(AuthProvider):
    """Auth provider based on a local storage of users in Home Assistant config dir."""

    DEFAULT_TITLE = "Home Assistant Local"

    def __init__(self, *args: Any, **kwargs: Any) -> None:
        """Initialize an Home Assistant auth provider."""
        super().__init__(*args, **kwargs)
        self.data: Data | None = None
        self._init_lock = asyncio.Lock()

    async def async_initialize(self) -> None:
        """Initialize the auth provider."""
        async with self._init_lock:
            if self.data is not None:
                return

            data = Data(self.hass)
            await data.async_load()
            self.data = data

    async def async_login_flow(self, context: dict[str, Any] | None) -> LoginFlow:
        """Return a flow to login."""
        return HassLoginFlow(self)

    async def async_validate_login(self, username: str, password: str) -> None:
        """Validate a username and password."""
        if self.data is None:
            await self.async_initialize()
            assert self.data is not None

        await self.hass.async_add_executor_job(
            self.data.validate_login, username, password
        )

    async def async_add_auth(self, username: str, password: str) -> None:
        """Call add_auth on data."""
        if self.data is None:
            await self.async_initialize()
            assert self.data is not None

        await self.hass.async_add_executor_job(self.data.add_auth, username, password)
        await self.data.async_save()

    async def async_remove_auth(self, username: str) -> None:
        """Call remove_auth on data."""
        if self.data is None:
            await self.async_initialize()
            assert self.data is not None

        self.data.async_remove_auth(username)
        await self.data.async_save()

    async def async_change_password(self, username: str, new_password: str) -> None:
        """Call change_password on data."""
        if self.data is None:
            await self.async_initialize()
            assert self.data is not None

        await self.hass.async_add_executor_job(
            self.data.change_password, username, new_password
        )
        await self.data.async_save()

    async def async_get_or_create_credentials(
        self, flow_result: Mapping[str, str]
    ) -> Credentials:
        """Get credentials based on the flow result."""
        if self.data is None:
            await self.async_initialize()
            assert self.data is not None

        norm_username = self.data.normalize_username
        username = norm_username(flow_result["username"])

        for credential in await self.async_credentials():
            if norm_username(credential.data["username"]) == username:
                return credential

        # Create new credentials.
        return self.async_create_credentials({"username": username})

    async def async_user_meta_for_credentials(
        self, credentials: Credentials
    ) -> UserMeta:
        """Get extra info for this credential."""
        return UserMeta(name=credentials.data["username"], is_active=True)

    async def async_will_remove_credentials(self, credentials: Credentials) -> None:
        """When credentials get removed, also remove the auth."""
        if self.data is None:
            await self.async_initialize()
            assert self.data is not None

        try:
            self.data.async_remove_auth(credentials.data["username"])
            await self.data.async_save()
        except InvalidUser:
            # Can happen if somehow we didn't clean up a credential
            pass


class HassLoginFlow(LoginFlow):
    """Handler for the login flow."""

    async def async_step_init(
        self, user_input: dict[str, str] | None = None
    ) -> FlowResult:
        """Handle the step of the form."""
        errors = {}

        if user_input is not None:
            try:
                await cast(HassAuthProvider, self._auth_provider).async_validate_login(
                    user_input["username"], user_input["password"]
                )
            except InvalidAuth:
                errors["base"] = "invalid_auth"

            if not errors:
                user_input.pop("password")
                return await self.async_finish(user_input)

        return self.async_show_form(
            step_id="init",
            data_schema=vol.Schema(
                {
                    vol.Required("username"): str,
                    vol.Required("password"): str,
                }
            ),
            errors=errors,
        )
