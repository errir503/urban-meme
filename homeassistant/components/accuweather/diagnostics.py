"""Diagnostics support for AccuWeather."""
from __future__ import annotations

from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_API_KEY, CONF_LATITUDE, CONF_LONGITUDE
from homeassistant.core import HomeAssistant

from . import AccuWeatherDataUpdateCoordinator
from .const import DOMAIN

TO_REDACT = {CONF_API_KEY, CONF_LATITUDE, CONF_LONGITUDE}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    coordinator: AccuWeatherDataUpdateCoordinator = hass.data[DOMAIN][
        config_entry.entry_id
    ]

    diagnostics_data = {
        "config_entry_data": async_redact_data(dict(config_entry.data), TO_REDACT),
        "coordinator_data": coordinator.data,
    }

    return diagnostics_data
