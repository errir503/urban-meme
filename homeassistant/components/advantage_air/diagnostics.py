"""Provides diagnostics for Advantage Air."""
from __future__ import annotations

from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant

from .const import DOMAIN as ADVANTAGE_AIR_DOMAIN

TO_REDACT = [
    "dealerPhoneNumber",
    "latitude",
    "logoPIN",
    "longitude",
    "postCode",
    "rid",
    "deviceNames",
    "deviceIds",
    "deviceIdsV2",
    "backupId",
]


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    data = hass.data[ADVANTAGE_AIR_DOMAIN][config_entry.entry_id].coordinator.data

    # Return only the relevant children
    return {
        "aircons": data.get("aircons"),
        "myLights": data.get("myLights"),
        "myThings": data.get("myThings"),
        "system": async_redact_data(data["system"], TO_REDACT),
    }
