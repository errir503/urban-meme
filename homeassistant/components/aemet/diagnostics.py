"""Support for the AEMET OpenData diagnostics."""
from __future__ import annotations

from typing import Any

from aemet_opendata.const import AOD_COORDS

from homeassistant.components.diagnostics.util import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    CONF_API_KEY,
    CONF_LATITUDE,
    CONF_LONGITUDE,
    CONF_UNIQUE_ID,
)
from homeassistant.core import HomeAssistant

from .const import DOMAIN, ENTRY_WEATHER_COORDINATOR
from .coordinator import WeatherUpdateCoordinator

TO_REDACT_CONFIG = [
    CONF_API_KEY,
    CONF_LATITUDE,
    CONF_LONGITUDE,
    CONF_UNIQUE_ID,
]

TO_REDACT_COORD = [
    AOD_COORDS,
]


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    aemet_entry = hass.data[DOMAIN][config_entry.entry_id]
    coordinator: WeatherUpdateCoordinator = aemet_entry[ENTRY_WEATHER_COORDINATOR]

    return {
        "api_data": coordinator.aemet.raw_data(),
        "config_entry": async_redact_data(config_entry.as_dict(), TO_REDACT_CONFIG),
        "coord_data": async_redact_data(coordinator.data, TO_REDACT_COORD),
    }
