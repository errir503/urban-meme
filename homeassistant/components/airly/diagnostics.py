"""Diagnostics support for Airly."""
from __future__ import annotations

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    CONF_API_KEY,
    CONF_LATITUDE,
    CONF_LONGITUDE,
    CONF_UNIQUE_ID,
)
from homeassistant.core import HomeAssistant

from . import AirlyDataUpdateCoordinator
from .const import DOMAIN

TO_REDACT = {CONF_API_KEY, CONF_LATITUDE, CONF_LONGITUDE, CONF_UNIQUE_ID}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: ConfigEntry
) -> dict:
    """Return diagnostics for a config entry."""
    coordinator: AirlyDataUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]

    diagnostics_data = {
        "config_entry": async_redact_data(config_entry.as_dict(), TO_REDACT),
        "coordinator_data": coordinator.data,
    }

    return diagnostics_data
