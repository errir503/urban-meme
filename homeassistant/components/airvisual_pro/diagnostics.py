"""Support for AirVisual Pro diagnostics."""
from __future__ import annotations

from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_PASSWORD
from homeassistant.core import HomeAssistant

from . import AirVisualProData
from .const import DOMAIN

CONF_MAC_ADDRESS = "mac_address"
CONF_SERIAL_NUMBER = "serial_number"

TO_REDACT = {
    CONF_MAC_ADDRESS,
    CONF_PASSWORD,
    CONF_SERIAL_NUMBER,
}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    data: AirVisualProData = hass.data[DOMAIN][entry.entry_id]

    return async_redact_data(
        {
            "entry": entry.as_dict(),
            "data": data.coordinator.data,
        },
        TO_REDACT,
    )
