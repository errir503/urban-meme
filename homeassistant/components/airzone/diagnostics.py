"""Support for the Airzone diagnostics."""
from __future__ import annotations

from typing import Any

from aioairzone.const import API_MAC, AZD_MAC

from homeassistant.components.diagnostics.util import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_UNIQUE_ID
from homeassistant.core import HomeAssistant

from .const import DOMAIN
from .coordinator import AirzoneUpdateCoordinator

TO_REDACT_API = [
    API_MAC,
]

TO_REDACT_CONFIG = [
    CONF_UNIQUE_ID,
]

TO_REDACT_COORD = [
    AZD_MAC,
]


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    coordinator: AirzoneUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]

    return {
        "api_data": async_redact_data(coordinator.airzone.raw_data(), TO_REDACT_API),
        "config_entry": async_redact_data(config_entry.as_dict(), TO_REDACT_CONFIG),
        "coord_data": async_redact_data(coordinator.data, TO_REDACT_COORD),
    }
