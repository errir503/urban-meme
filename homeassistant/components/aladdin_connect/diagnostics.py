"""Diagnostics support for Aladdin Connect."""
from __future__ import annotations

from typing import Any

from AIOAladdinConnect import AladdinConnectClient

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant

from .const import DOMAIN

TO_REDACT = {"serial", "device_id"}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""

    acc: AladdinConnectClient = hass.data[DOMAIN][config_entry.entry_id]

    diagnostics_data = {
        "doors": async_redact_data(acc.doors, TO_REDACT),
    }

    return diagnostics_data
