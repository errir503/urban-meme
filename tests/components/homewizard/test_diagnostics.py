"""Tests for diagnostics data."""

import pytest
from syrupy.assertion import SnapshotAssertion

from homeassistant.core import HomeAssistant

from tests.common import MockConfigEntry
from tests.components.diagnostics import get_diagnostics_for_config_entry
from tests.typing import ClientSessionGenerator


@pytest.mark.parametrize(
    "device_fixture", ["device-HWE-P1.json", "device-HWE-SKT.json"]
)
async def test_diagnostics(
    hass: HomeAssistant,
    hass_client: ClientSessionGenerator,
    init_integration: MockConfigEntry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test diagnostics."""
    assert (
        await get_diagnostics_for_config_entry(hass, hass_client, init_integration)
        == snapshot
    )
