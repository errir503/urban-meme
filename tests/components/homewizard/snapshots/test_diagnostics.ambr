# serializer version: 1
# name: test_diagnostics
  dict({
    'data': dict({
      'data': dict({
        'active_current_l1_a': -4,
        'active_current_l2_a': 2,
        'active_current_l3_a': 0,
        'active_frequency_hz': 50,
        'active_liter_lpm': 12.345,
        'active_power_average_w': 123.0,
        'active_power_l1_w': -123,
        'active_power_l2_w': 456,
        'active_power_l3_w': 123.456,
        'active_power_w': -123,
        'active_tariff': 2,
        'active_voltage_l1_v': 230.111,
        'active_voltage_l2_v': 230.222,
        'active_voltage_l3_v': 230.333,
        'any_power_fail_count': 4,
        'external_devices': None,
        'gas_timestamp': '2021-03-14T11:22:33',
        'gas_unique_id': '**REDACTED**',
        'long_power_fail_count': 5,
        'meter_model': 'ISKRA  2M550T-101',
        'monthly_power_peak_timestamp': '2023-01-01T08:00:10',
        'monthly_power_peak_w': 1111.0,
        'smr_version': 50,
        'total_gas_m3': 1122.333,
        'total_liter_m3': 1234.567,
        'total_energy_export_kwh': 13086.777,
        'total_energy_export_t1_kwh': 4321.333,
        'total_energy_export_t2_kwh': 8765.444,
        'total_energy_export_t3_kwh': None,
        'total_energy_export_t4_kwh': None,
        'total_energy_import_kwh': 13779.338,
        'total_energy_import_t1_kwh': 10830.511,
        'total_energy_import_t2_kwh': 2948.827,
        'total_energy_import_t3_kwh': None,
        'total_energy_import_t4_kwh': None,
        'unique_meter_id': '**REDACTED**',
        'voltage_sag_l1_count': 1,
        'voltage_sag_l2_count': 2,
        'voltage_sag_l3_count': 3,
        'voltage_swell_l1_count': 4,
        'voltage_swell_l2_count': 5,
        'voltage_swell_l3_count': 6,
        'wifi_ssid': '**REDACTED**',
        'wifi_strength': 100,
      }),
      'device': dict({
        'api_version': 'v1',
        'firmware_version': '2.11',
        'product_name': 'P1 Meter',
        'product_type': 'HWE-SKT',
        'serial': '**REDACTED**',
      }),
      'state': dict({
        'brightness': 255,
        'power_on': True,
        'switch_lock': False,
      }),
      'system': dict({
        'cloud_enabled': True,
      }),
    }),
    'entry': dict({
      'ip_address': '**REDACTED**',
    }),
  })
# ---
# name: test_diagnostics[device-HWE-P1.json]
  dict({
    'data': dict({
      'data': dict({
        'active_current_l1_a': -4,
        'active_current_l2_a': 2,
        'active_current_l3_a': 0,
        'active_frequency_hz': 50,
        'active_liter_lpm': 12.345,
        'active_power_average_w': 123.0,
        'active_power_l1_w': -123,
        'active_power_l2_w': 456,
        'active_power_l3_w': 123.456,
        'active_power_w': -123,
        'active_tariff': 2,
        'active_voltage_l1_v': 230.111,
        'active_voltage_l2_v': 230.222,
        'active_voltage_l3_v': 230.333,
        'any_power_fail_count': 4,
        'external_devices': None,
        'gas_timestamp': '2021-03-14T11:22:33',
        'gas_unique_id': '**REDACTED**',
        'long_power_fail_count': 5,
        'meter_model': 'ISKRA  2M550T-101',
        'monthly_power_peak_timestamp': '2023-01-01T08:00:10',
        'monthly_power_peak_w': 1111.0,
        'smr_version': 50,
        'total_energy_export_kwh': 13086.777,
        'total_energy_export_t1_kwh': 4321.333,
        'total_energy_export_t2_kwh': 8765.444,
        'total_energy_export_t3_kwh': 8765.444,
        'total_energy_export_t4_kwh': 8765.444,
        'total_energy_import_kwh': 13779.338,
        'total_energy_import_t1_kwh': 10830.511,
        'total_energy_import_t2_kwh': 2948.827,
        'total_energy_import_t3_kwh': 2948.827,
        'total_energy_import_t4_kwh': 2948.827,
        'total_gas_m3': 1122.333,
        'total_liter_m3': 1234.567,
        'unique_meter_id': '**REDACTED**',
        'voltage_sag_l1_count': 1,
        'voltage_sag_l2_count': 2,
        'voltage_sag_l3_count': 3,
        'voltage_swell_l1_count': 4,
        'voltage_swell_l2_count': 5,
        'voltage_swell_l3_count': 6,
        'wifi_ssid': '**REDACTED**',
        'wifi_strength': 100,
      }),
      'device': dict({
        'api_version': 'v1',
        'firmware_version': '4.19',
        'product_name': 'P1 meter',
        'product_type': 'HWE-P1',
        'serial': '**REDACTED**',
      }),
      'state': None,
      'system': dict({
        'cloud_enabled': True,
      }),
    }),
    'entry': dict({
      'ip_address': '**REDACTED**',
      'product_name': 'Product name',
      'product_type': 'product_type',
      'serial': '**REDACTED**',
    }),
  })
# ---
# name: test_diagnostics[device-HWE-SKT.json]
  dict({
    'data': dict({
      'data': dict({
        'active_current_l1_a': -4,
        'active_current_l2_a': 2,
        'active_current_l3_a': 0,
        'active_frequency_hz': 50,
        'active_liter_lpm': 12.345,
        'active_power_average_w': 123.0,
        'active_power_l1_w': -123,
        'active_power_l2_w': 456,
        'active_power_l3_w': 123.456,
        'active_power_w': -123,
        'active_tariff': 2,
        'active_voltage_l1_v': 230.111,
        'active_voltage_l2_v': 230.222,
        'active_voltage_l3_v': 230.333,
        'any_power_fail_count': 4,
        'external_devices': None,
        'gas_timestamp': '2021-03-14T11:22:33',
        'gas_unique_id': '**REDACTED**',
        'long_power_fail_count': 5,
        'meter_model': 'ISKRA  2M550T-101',
        'monthly_power_peak_timestamp': '2023-01-01T08:00:10',
        'monthly_power_peak_w': 1111.0,
        'smr_version': 50,
        'total_energy_export_kwh': 13086.777,
        'total_energy_export_t1_kwh': 4321.333,
        'total_energy_export_t2_kwh': 8765.444,
        'total_energy_export_t3_kwh': 8765.444,
        'total_energy_export_t4_kwh': 8765.444,
        'total_energy_import_kwh': 13779.338,
        'total_energy_import_t1_kwh': 10830.511,
        'total_energy_import_t2_kwh': 2948.827,
        'total_energy_import_t3_kwh': 2948.827,
        'total_energy_import_t4_kwh': 2948.827,
        'total_gas_m3': 1122.333,
        'total_liter_m3': 1234.567,
        'unique_meter_id': '**REDACTED**',
        'voltage_sag_l1_count': 1,
        'voltage_sag_l2_count': 2,
        'voltage_sag_l3_count': 3,
        'voltage_swell_l1_count': 4,
        'voltage_swell_l2_count': 5,
        'voltage_swell_l3_count': 6,
        'wifi_ssid': '**REDACTED**',
        'wifi_strength': 100,
      }),
      'device': dict({
        'api_version': 'v1',
        'firmware_version': '3.03',
        'product_name': 'Energy Socket',
        'product_type': 'HWE-SKT',
        'serial': '**REDACTED**',
      }),
      'state': dict({
        'brightness': 255,
        'power_on': True,
        'switch_lock': False,
      }),
      'system': dict({
        'cloud_enabled': True,
      }),
    }),
    'entry': dict({
      'ip_address': '**REDACTED**',
      'product_name': 'Product name',
      'product_type': 'product_type',
      'serial': '**REDACTED**',
    }),
  })
# ---
